design/open/repository-operations-error-handling.md

```diff
@@ -0,0 +1,121 @@
+* A `unique_identifier` for each `Repository` that will allow to differ it from other (previously created and possibly deleted) `Repository` that had the same `RepositoryID`. This will allow to correlate entities to the correct `Repository` entity, in case of `RepositoryID` collision (due to entities left from a previously deleted `Repository` with the same `RepositoryID`)
+In addition:
+* A garbage cleaning mechanism will remove zombie entities, and maintain our KV Store clean and as compact as possible. This GC will identify useless `Repository` entities and will reattempt to delete all the correlated entities, which are now easily recognizable with the addition of the new `unique_identifier`. Note that this GC is not a part of maintaining the DB consistency
+
+### Repository State
+Introduce a new attribute - `repository_state` - to recognize a failed or otherwise irrelevant `Repository` entities. Based on the `state`, a `Repository` that was not completely created, can be identified and treated as `not_exist` as well as a `Repository` in the middle of deletion.
+These unused `Repository` entries, which now can be easily detected, can later be cleaned by a future garbage cleaner, to maintain the KV cleanliness and to reduce its size.
+The new `repository_state` attribute can have one of four values: `initial`, `active`, `deleted` or `failed`
+* Upon creation, a `Repository` entry will have an `initial` state, to indicate it is not usable yet. It will also indicate that this `Repository` is soon to be valid for usage, so other threads/processes treat it as such
+  * A failure in the next creation steps - default `Branch`, initial `Commit` or any other future step, will transfer the `Repository` to state `failed`, if possible. If that cannot be achieved (e.g. the creating process crashed, or KV connectivity is lost) the `Repository` can stay in state `initial` as it is still unused
+  * If during creation, a `Repository` with identical ID (name) is found, the creation of the new `Repository` is stopped before any update the the `KV` is done. This is similar to the current behavior and will be done regardless of the existing `Repository` status
+* When retrieving a `Repository` it will only be retrieved if it is in `active` status. Otherwise an error will be returned.
+  * If the `Repository` entry does not exist, an `ErrRepositoryNotFound` will be returned
+  * If the `Repository` entry exists and is not in `active` state - **TBD #1** - should we return `ErrRepositoryNotFound` here too? Maybe a new error - `ErrRepositoryDeletionInProgress`-ish?
+* Deleting a `Repository` will be done by first setting the `Repository` state to `deleted`, making it immediately unavailable. Once the `Repository` is set to deleted all its correlated entities can be removed safely, as none of them should be accessible. In case of a failure at any step, the `Repository` and all its entities remain unreachable due to the `deleted` status.
```

Comment:
Why do we actually need the deleted state? Once we move it to failed state - it is inaccessible and eligible for deletion

Reply:
The only thing I had in mind is the option to return a more targeted information to the user, to allow different handling. For example if `DeleteRepository` returns `ErrorNotFound` the user can try and create it, if it's `ErrDeleteInProgress` the user might wait and then try again, and for any other error the user can retry immediately or alike.

Reply:
Referencing the previous comment: What difference does it make for the user if the repository is currently being deleted? For all things considered this repo doesn't exist and the user should be able to create a new one with the same name immediately after calling delete